src/box/applier.cc

```diff
@@ -800,7 +798,9 @@ applier_parse_tx_row(struct applier_tx_row *tx_row)
 			diag_raise();
 		}
 	} else if (type == IPROTO_OK) {
-		xrow_decode_heartbeat_xc(row, NULL, &tx_row->req.vclock_sync);
+		struct downstream_ack_request last_recv_ack;
+		xrow_decode_downstream_ack_xc(row, &last_recv_ack);
```

Comment:
Maybe switch to `struct relay_ack_request` and `struct applier_ack_request`?

Why does an applier decode a "downstream ack" when this ack came from the upstream?
IMO "relay ack" and "applier ack" would be simpler to understand. You even name the 'upstream ack' a 'downstream ack' by mistake in the last commit's mesage.

Reply:
Or probably they should be 'relay_heartbeat' and 'applier_heartbeat'. Because relay doesn't have acks.

Reply:
> Why does an applier decode a "downstream ack" when this ack came from the upstream?

It comes from downstream (`relay -> applier` is downstream).

> IMO "relay ack" and "applier ack" would be simpler to understand.

Apparently it is not much better. Besides, it is unclear whether `applier_ack` comes from applier or to applier?

> You even name the 'upstream ack' a 'downstream ack' by mistake in the last commit's mesage

It was a typo. Thanks for noticing, fixed.

> Or probably they should be 'relay_heartbeat' and 'applier_heartbeat'. Because relay doesn't have acks.

On the other hand applier doesn't have heartbeats. Anyway, I renamed them both to `..._heartbeat`.

Reply:
> It comes from downstream (relay -> applier is downstream).

It's a downstream from the master's point of view. But for a replica, it's an upstream. And replica sends ACKs to upstream, and receives heartbeats from the upstream. So yeah, this is complicated regardless of the name we choose.

Anyway, thanks for the fixes!